json_object.c

```diff
@@ -1138,60 +1138,79 @@ static void json_object_string_delete(struct json_object *jso)
 	json_object_generic_delete(jso);
 }
 
-struct json_object *json_object_new_string(const char *s)
+struct json_object *json_object_new_string_ext(const char *s, const int len,
+                                               int flags)
 {
-	struct json_object *jso = json_object_new(json_type_string);
-	if (!jso)
+	if (len <= 0 || flags == 0)
 		return NULL;
-	jso->_delete = &json_object_string_delete;
-	jso->_to_json_string = &json_object_string_to_json_string;
-	jso->o.c_string.len = strlen(s);
-	if (jso->o.c_string.len < LEN_DIRECT_STRING_DATA)
-	{
-		memcpy(jso->o.c_string.str.data, s, jso->o.c_string.len);
-	}
-	else
-	{
-		jso->o.c_string.str.ptr = strdup(s);
-		if (!jso->o.c_string.str.ptr)
-		{
-			json_object_generic_delete(jso);
-			errno = ENOMEM;
-			return NULL;
-		}
-	}
-	return jso;
-}
 
-struct json_object *json_object_new_string_len(const char *s, const int len)
-{
-	char *dstbuf;
 	struct json_object *jso = json_object_new(json_type_string);
 	if (!jso)
 		return NULL;
+	char *dstbuf = NULL;
 	jso->_delete = &json_object_string_delete;
 	jso->_to_json_string = &json_object_string_to_json_string;
+	jso->o.c_string.len = len;
+
+	/* short cut if the internal string buffer of jso is big enough */
 	if (len < LEN_DIRECT_STRING_DATA)
 	{
 		dstbuf = jso->o.c_string.str.data;
+		memcpy(dstbuf, s, jso->o.c_string.len);
+		dstbuf[len] = '\0';
+		if (flags & JSON_C_NEW_STRING_NO_STRDUP)
+		{
+			/* a little hack to get rid of constness */
+			memcpy(&dstbuf, &s, sizeof(char *));
+			/* string buffer must be free'd here
+			   to avoid memory leaks */
+			free(dstbuf);
```

Comment:
To actually support this, we'll need a different way to indicate whether the string is stored directly or not.  Perhaps, since c_string.len is signed, we could use negative lengths to mean it's stored directly?  hmm...

Reply:
> We can't guarantee that calling free() will be valid. Just because you don't want to strdup the string doesn't mean that it's memory from malloc. It might be global memory, or local variables that you happen to know won't go out of scope before the jso does, or a chunk of memory from an allocator other than malloc.

Then we would hit the same problem when the jso gets destroyed, which would call `free()` on the string buffer through `json_object_string_delete()` as well…

Reply:
I've marked all changes, that I have made localy, as resolved to keep track of them.  I'll push them as soon as we have found a solution here.

Reply:
Well sure, so the code in json_object_string_delete would need to do `if (jso->o.c_string.len < 0)` instead of `if (jso->o.c_string.len >= LEN_DIRECT_STRING_DATA)`, and every other place that currently references LEN_DIRECT_STRING_DATA would need some kind of change.  There are 7 of those total.

Reply:
Wouldn't it be better to have a parameter which specifies a func ptr to the `free()` function that corresponds to the undelying memory-allocator? And maybe `NULL` to use the system default `free()`? For stack-allocated objects we could offer a noop-function as well…

That seems better than relying on users to specify a negative value for the buffer length…

Reply:
You misunderstand me.  I'm not saying that the caller would specify a negative value, it would be the json-c code that would use a negative value if _it_ decides to store the string directly.

Reply:
...and in the no-strdup case we would _never_ copy the string into direct storage.

Reply:
i.e.:
```
...
if ( !(flags & JSON_C_NEW_STRING_NO_STRDUP) &&
     (len < LEN_DIRECT_STRING_DATA))
{
    ...
    jso->o.c_string.len = -len;
    return jso;
}
...
```